Userland/Libraries/LibGfx/ImageFormats/JPEGLoader.cpp

```diff
@@ -1272,6 +1272,97 @@ static void ycbcr_to_rgb(JPEGLoadingContext const& context, Vector<Macroblock>&
+{
+    if (!context.color_transform.has_value())
+        return;
+
+    // From libjpeg-turbo's libjpeg.txt:
+    // https://github.com/libjpeg-turbo/libjpeg-turbo/blob/main/libjpeg.txt
+    // CAUTION: it appears that Adobe Photoshop writes inverted data in CMYK JPEG
+    // files: 0 represents 100% ink coverage, rather than 0% ink as you'd expect.
+    // This is arguably a bug in Photoshop, but if you need to work with Photoshop
+    // CMYK files, you will have to deal with it in your application.
+    for (u32 vcursor = 0; vcursor < context.mblock_meta.vcount; vcursor += context.vsample_factor) {
+        for (u32 hcursor = 0; hcursor < context.mblock_meta.hcount; hcursor += context.hsample_factor) {
+            for (u8 vfactor_i = 0; vfactor_i < context.vsample_factor; ++vfactor_i) {
+                for (u8 hfactor_i = 0; hfactor_i < context.hsample_factor; ++hfactor_i) {
+                    u32 mb_index = (vcursor + vfactor_i) * context.mblock_meta.hpadded_count + (hcursor + hfactor_i);
+                    for (u8 i = 0; i < 8; ++i) {
+                        for (u8 j = 0; j < 8; ++j) {
+                            macroblocks[mb_index].r[i * 8 + j] = NumericLimits<u8>::max() - macroblocks[mb_index].r[i * 8 + j];
+                            macroblocks[mb_index].g[i * 8 + j] = NumericLimits<u8>::max() - macroblocks[mb_index].g[i * 8 + j];
+                            macroblocks[mb_index].b[i * 8 + j] = NumericLimits<u8>::max() - macroblocks[mb_index].b[i * 8 + j];
+                            macroblocks[mb_index].k[i * 8 + j] = NumericLimits<u8>::max() - macroblocks[mb_index].k[i * 8 + j];
+                        }
+                    }
+                }
+            }
+        }
+    }
+}
+
+static void cmyk_to_rgb(JPEGLoadingContext const& context, Vector<Macroblock>& macroblocks)
+{
+    invert_colors_for_adobe_images(context, macroblocks);
+
+    for (u32 vcursor = 0; vcursor < context.mblock_meta.vcount; vcursor += context.vsample_factor) {
+        for (u32 hcursor = 0; hcursor < context.mblock_meta.hcount; hcursor += context.hsample_factor) {
+            for (u8 vfactor_i = context.vsample_factor - 1; vfactor_i < context.vsample_factor; --vfactor_i) {
+                for (u8 hfactor_i = context.hsample_factor - 1; hfactor_i < context.hsample_factor; --hfactor_i) {
+                    u32 mb_index = (vcursor + vfactor_i) * context.mblock_meta.hpadded_count + (hcursor + hfactor_i);
+                    i32* c = macroblocks[mb_index].y;
+                    i32* m = macroblocks[mb_index].cb;
+                    i32* y = macroblocks[mb_index].cr;
+                    i32* k = macroblocks[mb_index].k;
+                    for (u8 i = 7; i < 8; --i) {
+                        for (u8 j = 7; j < 8; --j) {
+                            u8 const pixel = i * 8 + j;
+                            auto const black_component = 255 - k[pixel];
```

Comment:
Just an observation: you're using both `NumericLimits<u8>::max()` and `255`, what's the reason for one or the other? I don't mind either version but it's not really consistent :^)

Reply:
Looks like I copied this loop without updating it :sweat_smile:
I also removed the UB in the two most inner loops (we are now counting in ascending order).